Add card action effects

oliverlee opened the pull request:
TODO
- [x] Fix card drawing
- [x] Rename card resources (worth/wealth -> copper)
- [x] Add arena methods for moving cards, playing action cards (without consuming resources)
- [x] Add action stack
- [ ] ~Add arena method for revealing reactions~

oliverlee committed:
Split Arena::select_card into multiple simpler methods

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -115,88 +119,103 @@ impl Arena {
 
+        let card_index = self.players[player_id]
+            .hand
+            .iter()
```
This method is doing two things: 
1) find the first index of a card with the given kind in a players' hand.
2) play the gard at the given index.

I can imagine 1 being useful when writing bots but at other times we will already know the index and just need 2.

mickvangelderen replied:
I suggest to put 1 as a convenience method of the player or leave it out of the core library and just take a card hand index in this method.

oliverlee replied:
I think you're right about there being situations where we will already know the index and just need to play the card at the given index but I'll put off changing this until that situation comes up.

mickvangelderen replied:
In all of the test cases you know the card index because you just push it into the hand. 

Lets say you're writing a bot and you want it to play two cards if they're both in the hand. You can't do this with the current API without searching through your hands unnecessarily.

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -115,88 +119,103 @@ impl Arena {
+            .position(|&x| x == card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        let gains = card.action().ok_or_else(|| Error::InvalidCardChoice)?;
```
move this into the match, closer to where its used

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -115,88 +119,103 @@ impl Arena {
+        }
+    }
+
+    pub fn buy_card(&mut self, player_id: usize, card: CardKind) -> Result<()> {
```
Why would we ever want to call this with a player_id != current player id? It doesn't even work because of the assert in that case. Either remove the assertion or the argument.

mickvangelderen replied:
Removing the assertion means that we'll be using the wrong turn phase. So remove the player_id parameter.

oliverlee replied:
When we implement the server/client functionality, it is possible that a player will try and to buy out of turn. We'll need some way to allow some functions (e.g. view the cards in the trash) without other (e.g. buy a card).

If we remove `player_id ` here, we'll need to validate that the client/player is the current one.

mickvangelderen replied:
We'll need to validate a whole bunch of things because we can't trust the input. This does not protect against another player trying to force the current player to buy a curse card. 

The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

oliverlee replied:
> The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

I agree this makes sense and can see that the function `play_action`/`buy_card`/etc. would be simpler without checking ` player_id`. However, I don't want remove `player_id` as a parameter until the privilege checking is implemented and I want to test out my effect stack idea first =p

mickvangelderen replied:
Then debug_assert!(...) is more appropriate

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -115,88 +119,103 @@ impl Arena {
+            .get_mut(card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        match &mut self.turn.phase {
```
This turn phase thing happens often, you can abstract it
```rust
impl TurnPhase {
fn as_buy_phase_mut(&mut self) -> Result<(&mut remaining_buys, &mut total_wealth), Error::WrongTurnPhase> {
match self {
    TurnPhase::Buy { remaining_buys, total_wealth } => Ok((remaining_buys, total_wealth)),
    _ => Err(Error::WrongTurnPhase),
}
}

Instead of the tuple you can have a struct `BuyPhase` with the fields and change the variants to `Buy(BuyPhase)` etc.
```

mickvangelderen commented on dominion/src/main.rs:
```diff
@@ -56,9 +56,25 @@ fn big_money(arena: &mut Arena, player_id: usize) {
+        })
+        .unwrap();
 
     arena.end_buy_phase(player_id).unwrap();
```
Since we keep track of the current state (player, turn) and all the other methods check if the game is in the right state, this function can simply be `end_phase()` which instead of erroring just does the right thing?

oliverlee replied:
We would still need to return a result (in case a non-current player) tries to end the current turn phase. I think it would be better to explicitly use `end_action_phase` and `end_buy_phase` in the case that a user sends two `end_phase` commands by mistake somehow.

mickvangelderen reviewed (changes requested):
Some suggestions.

oliverlee committed:
Define ActionPhase and BuyPhase for TurnPhase enums

Define ActionPhase and BuyPhase structs to abstract use of TurnPhase
resources. Rename wealth/worth to copper.

oliverlee committed:
Add Arena move card method

oliverlee committed:
Add initial ActionEffectStack implementation

The ActionEffectStack is implemented as a stack of functions that must
be resolved to complete play of an Action card. A stack consists of
three types of items:
- Source: Denotes that a specific player must make a card selection
  according to a condition.
- Sink: A functions that modifies the Arena state in reaction to a
  previous card selection of a previous Source.
- Function: A function that modifies the Arena state, generally for a
  specific player.

When resolving an ActionEffectStack, the same parameters are used to
evaluate all items until a Sink condition is not satisfied.

This commit defines the ActionEffectStack for:
- Market
- Militia

Note that a BuyEffectStack will be necessary for some Action cards (i.e.
Merchant).

oliverlee committed:
Allow ActionEffect to spawn additional effects

Define ActionEffect to contain ActionCardEffect`s. Return an Option of
cards when resolving an ActionEffectFunc and if not None, append
additional effects for each card.

Define ActionEffect and tests for ThroneRoom.

oliverlee committed:
Don't panic if a card cannot be drawn

oliverlee committed:
Refactor Arena methods to use Arena::move_card

oliverlee committed:
Block Arena mutation if an action effect is unresolved

This commit also removes player_id as an input paramter from mutable
Arena methods as they can only be used by the current player. Validation
of inputs will need to be performed separately.

andebjor commented on dominion/src/dominion/arena.rs:
```diff
@@ -370,154 +457,154 @@ mod tests {
     fn buy_card_not_enough_wealth() {
```
Rename to `buy_card_not_enough_copper`?

andebjor commented on dominion/src/dominion/player.rs:
```diff
@@ -85,44 +86,44 @@ mod tests {
+        assert_eq!(p.draw_pile, vec![CardKind::Copper; 4]);
         assert_eq!(p.hand, vec![CardKind::Copper]);
     }
 
```
Suggest adding a test with an empty `draw_pile` and empty `discard_pile`. Should work now, right?

oliverlee committed:
Simply card action effects

Rename {ActionEffect,ActionCardEffect,ActionEffectFunc} to
{CardActionQueue,CardAction,Effect}. Combine Source/Sink enum variants
so that there are only two Effect enum variants: Conditional and
Unconditional. Effect::Conditional requires player input and
Effect::Unconditional does not. Effect::Conditional will return an error
if player input does not satisfy a specified condition.

mickvangelderen reviewed (commented):
Just as a remark, pub(crate) fields are useful if there are no invariants to uphold but you don't want to make the field part of the external API. If there's no invariants to uphold, the field might as well be pub. If there are invariants to uphold, perhaps it is better to expose pub(crate) methods that will do so. I am of the opinion that  they shouldn't be pub(crate) out of lazyness, even in early development, because it creates more opportunities to break invariants. If a bunch of types work closely together then it is reasonable to put the in the same module so private fields can be accessed.

If you have anything specific you'd like me to review let me know, otherwise I will simply approve the changes for the sake of moving forwards.

oliverlee commented:
👍 We will probably want to restructure the code and module privacy at some
point as most things should be hidden from the public API (I'm guessing
that we'll probably only want arena methods and card types visible).

On Thu, Jun 13, 2019, 11:51 Mick van Gelderen <notifications@github.com>
wrote:

> *@mickvangelderen* commented on this pull request.
>
> Just as a remark, pub(crate) fields are useful if there are no invariants
> to uphold but you don't want to make the field part of the external API. If
> there's no invariants to uphold, the field might as well be pub. If there
> are invariants to uphold, perhaps it is better to expose pub(crate) methods
> that will do so. I am of the opinion that they shouldn't be pub(crate) out
> of lazyness, even in early development, because it creates more
> opportunities to break invariants. If a bunch of types work closely
> together then it is reasonable to put the in the same module so private
> fields can be accessed.
>
> If you have anything specific you'd like me to review let me know,
> otherwise I will simply approve the changes for the sake of moving forwards.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/oliverlee/dominion/pull/13?email_source=notifications&email_token=AAWT4J66DDQ2D7RQJBGF7BDP2IKDHA5CNFSM4HWG7WDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3NR6WI#pullrequestreview-249241433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWT4J4FSMV3S2MPJGJMGVTP2IKDHANCNFSM4HWG7WDA>
> .
>

mickvangelderen reviewed (dismissed):
Lets move forward

oliverlee commented:
I'd like to clean up the commit messages before merging this.

On Thu, Jun 13, 2019 at 1:57 PM Mick van Gelderen <notifications@github.com>
wrote:

> *@mickvangelderen* approved this pull request.
>
> Lets move forward
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/oliverlee/dominion/pull/13?email_source=notifications&email_token=AAWT4J5FZBWPAT2M4EQRVI3P2IY3VA5CNFSM4HWG7WDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3N7VPQ#pullrequestreview-249297598>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWT4J5NL5TMIUFUPRWOAKDP2IY3VANCNFSM4HWG7WDA>
> .
>

oliverlee commented:
Although it shouldn't stop anyone from branching off of this commit.

On Thu, Jun 13, 2019 at 2:01 PM Oliver Lee <oliverzlee@gmail.com> wrote:

> I'd like to clean up the commit messages before merging this.
>
> On Thu, Jun 13, 2019 at 1:57 PM Mick van Gelderen <
> notifications@github.com> wrote:
>
>> *@mickvangelderen* approved this pull request.
>>
>> Lets move forward
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/oliverlee/dominion/pull/13?email_source=notifications&email_token=AAWT4J5FZBWPAT2M4EQRVI3P2IY3VA5CNFSM4HWG7WDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3N7VPQ#pullrequestreview-249297598>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAWT4J5NL5TMIUFUPRWOAKDP2IY3VANCNFSM4HWG7WDA>
>> .
>>
>
